priv/rest_xml.ex.eex

```diff
@@ -0,0 +1,147 @@
+          |> Enum.reduce(body, fn {header_name, key}, acc ->
+            case List.keyfind(response.headers, header_name, 0) do
+              nil -> acc
+              {_header_name, value} -> Map.put(acc, key, value)
+            end
+          end)
+
+        {:ok, body, response}
+
+      result ->
+        result
+    end<% else %>
+    request(client, <%= AWS.CodeGen.RestService.Action.method(action) %>, path, headers, input, options, <%= inspect(action.success_status_code) %>)<% end %>
+  end<% end %><% end %>
+
+  @spec request(AWS.Client.t(), binary(), binary(), list(), map(), list(), pos_integer()) ::
+          {:ok, Poison.Parser.t() | nil, Poison.Response.t()}
+          | {:error, Poison.Parser.t()}
+          | {:error, HTTPoison.Error.t()}
+  defp request(client, method, path, headers, input, options, success_status_code) do
+    client = %{client | service: "<%= context.signing_name %>"<%= if context.is_global do %>,
+                        region:  "<%= context.credential_scope %>"<% end %>}
+    host = get_host("<%= context.endpoint_prefix %>", client)
+    url = get_url(host, path, client)
+
+    headers = [
+      {"Host", host},
+      {"Content-Type", "text/xml"}
+      | headers
+    ]
+
+    payload = encode_payload(input)
+    headers = AWS.Request.sign_v4(client, method, url, headers, payload)
+    perform_request(method, url, payload, headers, options, success_status_code)
+  end
+
+  defp perform_request(method, url, payload, headers, options, nil) do
+    case HTTPoison.request(method, url, payload, headers, options) do
+      {:ok, %HTTPoison.Response{status_code: 200, body: ""} = response} ->
+        {:ok, response}
+
+      {:ok, %HTTPoison.Response{status_code: status_code, body: body} = response}
+      when status_code == 200 or status_code == 202 or status_code == 204 ->
+        {:ok, AWS.Util.decode_xml(body), response}
+
+      {:ok, %HTTPoison.Response{body: body}} ->
+        error = AWS.Util.decode_xml(body)
+        reason = error["ErrorResponse"]["Error"]["Message"]
+        {:error, reason}
+
+      {:error, %HTTPoison.Error{reason: reason}} ->
+        {:error, %HTTPoison.Error{reason: reason}}
+    end
+  end
+
+  defp perform_request(method, url, payload, headers, options, success_status_code) do
+    case HTTPoison.request(method, url, payload, headers, options) do
+      {:ok, %HTTPoison.Response{status_code: ^success_status_code, body: ""} = response} ->
+        {:ok, nil, response}
+
+      {:ok, %HTTPoison.Response{status_code: ^success_status_code, body: body} = response} ->
+        {:ok, AWS.Util.decode_xml(body), response}
+
+      {:ok, %HTTPoison.Response{body: body}} ->
+        error = AWS.Util.decode_xml(body)
+        reason = error["ErrorResponse"]["Error"]["Message"]
+        {:error, reason}
+
+      {:error, %HTTPoison.Error{reason: reason}} ->
+        {:error, %HTTPoison.Error{reason: reason}}
+    end
+  end
+
+  defp get_host(_endpoint_prefix, %{region: "local"}) do
```

Comment:
We should handle the special case of `s3-control` in the Elixir template as well.